File: py/lexer.py
from typing import List, Union

Token = Union[str, int]


class Lexer:
    def __init__(self, source: str):
        self.source = source
        self.rawTokens = []

    @property
    def lastRawToken(self):
        return self.rawTokens[-1] if self.rawTokens else ""

    @lastRawToken.setter
    def lastRawToken(self, token: str):
        if self.rawTokens:
            self.rawTokens[-1] = token

    def process(self, char: str):
        if char == "&" or char == "|":
            if self.lastRawToken == char:
                self.lastRawToken = char + char
            else:
                self.rawTokens.append(char)
        elif char == " " or char == "\t":
            if self.lastRawToken != " ":
                self.rawTokens.append(" ")
        elif char == "=":
            last = self.lastRawToken
            if last == ">" or last == "<" or last == "=" or last == "!":
                self.lastRawToken = last + char
            else:
                self.rawTokens.append(char)
        elif (
            char == "+" or
            char == "-" or
            char == "*" or
            char == "/" or
            char == "#" or
            char == "!" or
            char == ">" or
            char == "<"
        ):
            self.rawTokens.append(char)
        else:
            if self.lastRawToken.isalnum():
                self.lastRawToken = self.lastRawToken + char
            else:
                self.rawTokens.append(char)

    def run(self, source: str):
        for c in source:
            self.process(c)

    @property
    def tokens(self) -> List[Token]:
        return [
            int(v) if str(v).isdigit() else v for v in map(lambda v:'&' if v == 'in' else v, self.rawTokens) if v != " "
        ]

    @staticmethod
    def tokenlize_line(line: str) -> List[Token]:
        lexer = Lexer(line)
        lexer.run(line)
        return lexer.tokens

File: js/src/lexer.ts
export type Token = string | number;

class CommentException {}

export class Lexer {
  public rawTokens: string[] = [];
  public comment = null as string | null;
  protected get lastRawToken() {
    return this.rawTokens[this.rawTokens.length - 1] ?? "";
  }
  protected set lastRawToken(token: string) {
    this.rawTokens[this.rawTokens.length - 1] = token;
  }
  protected process(char: string) {
    if (char.trim().length === 0) {
      if (
        this.lastRawToken.trim().length === 0 &&
        this.lastRawToken !== "" /**not the first token */
      ) {
        this.lastRawToken = this.lastRawToken + char;
      } else {
        this.rawTokens.push(char);
      }
    } else if (char === "#") {
      throw new CommentException();
    } else if (char === "&" || char === "|") {
      if (this.lastRawToken === char) {
        this.lastRawToken = char + char;
      } else {
        this.rawTokens.push(char);
      }
    } else if (char === "=") {
      const last = this.rawTokens.at(-1);
      if (last === ">" || last === "<" || last === "=" || last === "!") {
        this.lastRawToken = last + char;
      } else {
        this.rawTokens.push(char);
      }
    } else if (
      char === "+" ||
      char === "-" ||
      char === "*" ||
      char === "/" ||
      char === "#" ||
      char === "!" ||
      char === ">" ||
      char === "<"
    ) {
      this.rawTokens.push(char);
    } else {
      if (/^[a-zA-Z0-9]+$/.test(this.lastRawToken)) {
        this.lastRawToken = this.lastRawToken + char;
      } else {
        this.rawTokens.push(char);
      }
    }
  }

  public run(source: string) {
    for (let i = 0; i < source.length; i++) {
      const c = source[i];
      try {
        this.process(c);
      } catch (e) {
        if (e instanceof CommentException) {
          this.comment = source.slice(i + 1);
          break;
        } else {
          throw e;
        }
      }
    }
  }

  public get tokens(): Token[] {
    return this.rawTokens
      .filter((v) => v.trim().length > 0) // delete spaces
      .map((v) => (Number.isFinite(+v) ? +v : v)); // convert numeric to number type
  }

  public static tokenlizeLine(line: string) {
    const lexer = new Lexer();
    lexer.run(line);
    return [lexer.tokens, lexer.comment] as const;
  }
}
